### PR TITLE
feat: move selectKey impl to plan builder

### PR DESCRIPTION
**Pull request by rodesai**
### Description 
Moves rekeying from SchemaKStream and into an execution plan builder.

### Testing done 
Added tests for step builder

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-streams/src/test/java/io/confluent/ksql/execution/streams/StreamSelectKeyBuilderTest.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldRekeyCorrectly() {
+    // When:
+    final KStream result = StreamSelectKeyBuilder.build(kstream, selectKey);
+
+    // Then:
+    verify(kstream).filter(any());
```
Comment (big-andy-coates) [inline ksql-streams/src/test/java/io/confluent/ksql/execution/streams/StreamSelectKeyBuilderTest.java]: These should be in order though right?   Consider using mockito's `InOrder` to test this.

```java
final InOrder inOrder = inOrder(kstream, filteredKStream, rekeyedKstream);
inOrder.verify(kstream).filter(any());
inOrder.verify(filteredKStream).selectKey(any());
inOrder.verify(rekeyedKstream).mapValues(any(ValueMapperWithKey.class));
```

File: ksql-streams/src/test/java/io/confluent/ksql/execution/streams/StreamSelectKeyBuilderTest.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    verify(rekeyedKstream).mapValues(mapperCaptor.capture());
+    final ValueMapperWithKey<Object, GenericRow, GenericRow> mapper = mapperCaptor.getValue();
+    assertThat(
+        mapper.apply(asStructKey("bob"), new GenericRow(0, "dre", 3000, "bob")),
+        equalTo(new GenericRow(0, "bob", 3000, "bob"))
+    );
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
```
Comment (big-andy-coates) [inline ksql-streams/src/test/java/io/confluent/ksql/execution/streams/StreamSelectKeyBuilderTest.java]: consider moving to class scope?

File: ksql-streams/src/test/java/io/confluent/ksql/execution/streams/StreamSelectKeyBuilderTest.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+  public void shouldUpdateRowkeyIfUpdateRowkeyTrue() {
+    // When:
+    StreamSelectKeyBuilder.build(kstream, selectKey);
+
+    // Then:
+    verify(rekeyedKstream).mapValues(mapperCaptor.capture());
```
Comment (big-andy-coates) [inline ksql-streams/src/test/java/io/confluent/ksql/execution/streams/StreamSelectKeyBuilderTest.java]: consider moving into method to make tests more readable:

```java
    private ValueMapperWithKey<Object, GenericRow, GenericRow> getMapper() { 
       verify(rekeyedKstream).mapValues(mapperCaptor.capture());
       final ValueMapperWithKey<Object, GenericRow, GenericRow> mapper = mapperCaptor.getValue();
    }
```

Same for others.

File: ksql-streams/src/main/java/io/confluent/ksql/execution/streams/ExecutionStepFactory.java [hunk 4]
```diff
@@ -215,14 +215,15 @@ private ExecutionStepFactory() {
...
+  @SuppressWarnings("unchecked")
+  public static <K> StreamSelectKey<KStream<Struct, GenericRow>> streamSelectKey(
       final QueryContext.Stacker stacker,
       final ExecutionStep<KStream<K, GenericRow>> source,
       final String fieldName,
       final boolean updateRowKey
   ) {
     final QueryContext queryContext = stacker.getQueryContext();
-    return new StreamSelectKey<>(
+    return (StreamSelectKey<KStream<Struct, GenericRow>>) new StreamSelectKey(
```
Comment (big-andy-coates) [inline ksql-streams/src/main/java/io/confluent/ksql/execution/streams/ExecutionStepFactory.java]: use of raw type.

We need to get the generics right here.

Will `StreamSelectKey` always build a `KStream<Struct, GenericRow>` regardless of whether it's source is `KStream<Windowed<Struct>, GenericRow>` or `KStream<Struct, GenericRow>`??? I think it does, right?

That being the case, `StreamSelectKey` should be defined as:

```java
public class StreamSelectKey implements ExecutionStep<KStream<Struct, GenericRow>> {

  private final ExecutionStep<KStream<?, GenericRow>> source;
  ...

  public StreamSelectKey(
      ...
      final ExecutionStep<KStream<?, GenericRow>> source,
     ...
  )

  ...
}
```

And this method should be:

```java
public static StreamSelectKey streamSelectKey(
      final QueryContext.Stacker stacker,
      final ExecutionStep<KStream<?, GenericRow>> source,
      final String fieldName,
      final boolean updateRowKey
  ) {
    final QueryContext queryContext = stacker.getQueryContext();
    return new StreamSelectKey(
        new DefaultExecutionStepProperties(
            source.getProperties().getSchema(),
            queryContext
        ),
        source,
        fieldName,
        updateRowKey
    );
  }
```

File: ksql-streams/src/test/java/io/confluent/ksql/execution/streams/StreamSelectKeyBuilderTest.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+  private static final LogicalSchema SCHEMA = LogicalSchema.builder()
+      .valueColumn("BIG", SqlTypes.BIGINT)
+      .valueColumn("BOI", SqlTypes.STRING)
+      .build()
+      .withAlias(ALIAS)
+      .withMetaAndKeyColsInValue();
+  private static final String KEY = "ATL.BOI";
+
+  @Mock
+  private KStream kstream;
```
Comment (big-andy-coates) [inline ksql-streams/src/test/java/io/confluent/ksql/execution/streams/StreamSelectKeyBuilderTest.java]: Use of raw types - can we please use generic types in tests.

File: ksql-execution/src/main/java/io/confluent/ksql/execution/plan/StreamSelectKey.java [hunk 6]
```diff
@@ -15,21 +15,24 @@
...
 import io.confluent.ksql.execution.builder.KsqlQueryBuilder;
 import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.streams.kstream.KStream;
 
 @Immutable
-public class StreamSelectKey<S> implements ExecutionStep<S> {
+public class StreamSelectKey<K> implements ExecutionStep<KStream<Struct, GenericRow>> {
```
Comment (big-andy-coates) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/plan/StreamSelectKey.java]: Do we need `K` here?  Can we not just use `ExecutionStep<KStream<?, GenericRow>>` for the `source ` param / field

File: ksql-streams/src/test/java/io/confluent/ksql/execution/streams/StreamSelectKeyBuilderTest.java (hunk 3 above)
Comment (big-andy-coates) [reply ksql-streams/src/test/java/io/confluent/ksql/execution/streams/StreamSelectKeyBuilderTest.java]: consider moving into method to make tests more readable:

```java
    private ValueMapperWithKey<Object, GenericRow, GenericRow> getMapper() { 
       verify(rekeyedKstream).mapValues(mapperCaptor.capture());
       final ValueMapperWithKey<Object, GenericRow, GenericRow> mapper = mapperCaptor.getValue();
    }
```

Same for others.

File: ksql-execution/src/main/java/io/confluent/ksql/execution/plan/StreamSelectKey.java (hunk 6 above)
Comment (rodesai) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/plan/StreamSelectKey.java]: You can't pass an `ExecutionStep<KStream<K, GenericRow>>`, which is what `SchemaKStream` has, to `ExecutionStep<KStream<?, GenericRow>>`. We could make the type `ExecutionStep<? extends KStream<?, GenericRow>>`, but this felt easier to read.

**Pull request merged**